TestProcessSuite/runReaper=false/TestRunLogs

```
#177 39.28 === RUN   TestProcessSuite/runReaper=false/TestRunLogs
#177 39.28 2025/04/30 13:03:43 state Running: Process Process(["/bin/bash" "-c" "echo -n \"Test 1\nTest 2\n\""]) started with PID 12750
#177 39.28     process_test.go:96: 
#177 39.28         	Error Trace:	/src/internal/app/machined/pkg/system/runner/process/process_test.go:96
#177 39.28         	Error:      	Not equal: 
#177 39.28         	            	expected: []byte{0x54, 0x65, 0x73, 0x74, 0x20, 0x31, 0xa, 0x54, 0x65, 0x73, 0x74, 0x20, 0x32, 0xa}
#177 39.28         	            	actual  : []byte{}
#177 39.28         	            	
#177 39.28         	            	Diff:
#177 39.28         	            	--- Expected
#177 39.28         	            	+++ Actual
#177 39.28         	            	@@ -1,3 +1,2 @@
#177 39.28         	            	-([]uint8) (len=14) {
#177 39.28         	            	- 00000000  54 65 73 74 20 31 0a 54  65 73 74 20 32 0a        |Test 1.Test 2.|
#177 39.28         	            	+([]uint8) {
#177 39.28         	            	 }
#177 39.28         	Test:       	TestProcessSuite/runReaper=false/TestRunLogs
#177 39.28 --- FAIL: TestProcessSuite (2.44s)
#177 39.28     --- PASS: TestProcessSuite/runReaper=true (1.38s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestIOPriority (0.03s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestPriority (1.02s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestRunLogs (0.00s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestRunRestartFailed (0.10s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestRunSuccess (0.00s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestSchedulingPolicy (0.03s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestStopFailingAndRestarting (0.08s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=true/TestStopSigKill (0.11s)
#177 39.28     --- FAIL: TestProcessSuite/runReaper=false (1.06s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=false/TestIOPriority (0.03s)
#177 39.28         --- PASS: TestProcessSuite/runReaper=false/TestPriority (1.02s)
#177 39.28         --- FAIL: TestProcessSuite/runReaper=false/TestRunLogs (0.01s)
```

--- a/internal/app/machined/pkg/controllers/network/cmdline_test.go
+++ b/internal/app/machined/pkg/controllers/network/cmdline_test.go
@@ -23,8 +23,7 @@ import (
 )
 
 func TestCmdlineParse(t *testing.T) {
-	t.Parallel()
-
+	// [NOTE]: this test is not safe to run in parallel, as defaultIfaceName might flip if some interface is created concurrently
 	ifaces, _ := net.Interfaces() //nolint:errcheck // ignoring error here as ifaces will be empty
 
 	slices.SortFunc(ifaces, func(a, b net.Interface) int { return cmp.Compare(a.Name, b.Name) })
@@ -599,8 +598,6 @@ func TestCmdlineParse(t *testing.T) {
 		},
 	} {
 		t.Run(test.name, func(t *testing.T) {
-			t.Parallel()
-
 			cmdline := procfs.NewCmdline(test.cmdline)
 
 			link1 := netconfig.NewLinkStatus(netconfig.NamespaceName, "eth31")

--- a/internal/app/machined/pkg/system/runner/process/process.go
+++ b/internal/app/machined/pkg/system/runner/process/process.go
@@ -214,13 +214,17 @@ func (p *processRunner) build() (commandWrapper, error) {
 		return commandWrapper{}, err
 	}
 
-	go io.Copy(writer, pr) //nolint:errcheck
+	go func() {
+		defer pr.Close() //nolint:errcheck
+
+		io.Copy(writer, pr) //nolint:errcheck
+	}()
 
 	// close the writer if we exit early due to an error
 	closeWriter := true
 
 	closeLogging := func() (e error) {
-		for _, closer := range []io.Closer{w, pr, pw} {
+		for _, closer := range []io.Closer{w, pw} {
 			e = cmp.Or(closer.Close(), e)
 		}
 

--- a/internal/app/machined/pkg/system/runner/process/process_test.go
+++ b/internal/app/machined/pkg/system/runner/process/process_test.go
@@ -7,7 +7,6 @@ package process_test
 import (
 	"fmt"
 	"io"
-	"log"
 	"os"
 	"path/filepath"
 	"strconv"
@@ -18,6 +17,7 @@ import (
 	"time"
 
 	"github.com/siderolabs/go-cmd/pkg/cmd/proc/reaper"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/suite"
 
 	"github.com/siderolabs/talos/internal/app/machined/pkg/runtime"
@@ -28,8 +28,10 @@ import (
 	"github.com/siderolabs/talos/internal/app/machined/pkg/system/runner/restart"
 )
 
-func MockEventSink(state events.ServiceState, message string, args ...any) {
-	log.Printf("state %s: %s", state, fmt.Sprintf(message, args...))
+func MockEventSink(t *testing.T) func(state events.ServiceState, message string, args ...any) {
+	return func(state events.ServiceState, message string, args ...any) {
+		t.Logf("state %s: %s", state, fmt.Sprintf(message, args...))
+	}
 }
 
 type ProcessSuite struct {
@@ -67,7 +69,7 @@ func (suite *ProcessSuite) TestRunSuccess() {
 
 	defer func() { suite.Assert().NoError(r.Close()) }()
 
-	suite.Assert().NoError(r.Run(MockEventSink))
+	suite.Assert().NoError(r.Run(MockEventSink(suite.T())))
 	// calling stop when Run has finished is no-op
 	suite.Assert().NoError(r.Stop())
 }
@@ -82,18 +84,17 @@ func (suite *ProcessSuite) TestRunLogs() {
 
 	defer func() { suite.Assert().NoError(r.Close()) }()
 
-	suite.Assert().NoError(r.Run(MockEventSink))
+	suite.Assert().NoError(r.Run(MockEventSink(suite.T())))
 
-	logFile, err := os.Open(filepath.Join(suite.tmpDir, "logtest.log"))
-	suite.Assert().NoError(err)
+	// the log file is written asynchronously, so we need to wait a bit
+	suite.EventuallyWithT(func(collect *assert.CollectT) {
+		asrt := assert.New(collect)
 
-	//nolint:errcheck
-	defer logFile.Close()
+		logContents, err := os.ReadFile(filepath.Join(suite.tmpDir, "logtest.log"))
+		asrt.NoError(err)
 
-	logContents, err := io.ReadAll(logFile)
-	suite.Assert().NoError(err)
-
-	suite.Assert().Equal([]byte("Test 1\nTest 2\n"), logContents)
+		asrt.Equal([]byte("Test 1\nTest 2\n"), logContents)
+	}, time.Second, 10*time.Millisecond)
 }
 
 func (suite *ProcessSuite) TestRunRestartFailed() {
@@ -116,7 +117,7 @@ func (suite *ProcessSuite) TestRunRestartFailed() {
 
 	go func() {
 		defer wg.Done()
-		suite.Assert().NoError(r.Run(MockEventSink))
+		suite.Assert().NoError(r.Run(MockEventSink(suite.T())))
 	}()
 
 	fetchLog := func() []byte {
@@ -166,7 +167,7 @@ func (suite *ProcessSuite) TestStopFailingAndRestarting() {
 	done := make(chan error, 1)
 
 	go func() {
-		done <- r.Run(MockEventSink)
+		done <- r.Run(MockEventSink(suite.T()))
 	}()
 
 	time.Sleep(40 * time.Millisecond)
@@ -213,7 +214,7 @@ func (suite *ProcessSuite) TestStopSigKill() {
 	done := make(chan error, 1)
 
 	go func() {
-		done <- r.Run(MockEventSink)
+		done <- r.Run(MockEventSink(suite.T()))
 	}()
 
 	time.Sleep(100 * time.Millisecond)
@@ -253,7 +254,7 @@ func (suite *ProcessSuite) TestPriority() {
 	done := make(chan error, 1)
 
 	go func() {
-		done <- r.Run(MockEventSink)
+		done <- r.Run(MockEventSink(suite.T()))
 	}()
 
 	time.Sleep(10 * time.Millisecond)
@@ -307,7 +308,7 @@ func (suite *ProcessSuite) TestIOPriority() {
 	done := make(chan error, 1)
 
 	go func() {
-		done <- r.Run(MockEventSink)
+		done <- r.Run(MockEventSink(suite.T()))
 	}()
 
 	time.Sleep(10 * time.Millisecond)
@@ -360,7 +361,7 @@ func (suite *ProcessSuite) TestSchedulingPolicy() {
 	done := make(chan error, 1)
 
 	go func() {
-		done <- r.Run(MockEventSink)
+		done <- r.Run(MockEventSink(suite.T()))
 	}()
 
 	time.Sleep(10 * time.Millisecond)
@@ -383,8 +384,10 @@ func (suite *ProcessSuite) TestSchedulingPolicy() {
 
 func TestProcessSuite(t *testing.T) {
 	for _, runReaper := range []bool{true, false} {
-		func(runReaper bool) {
-			t.Run(fmt.Sprintf("runReaper=%v", runReaper), func(t *testing.T) { suite.Run(t, &ProcessSuite{runReaper: runReaper}) })
-		}(runReaper)
+		t.Run(fmt.Sprintf("runReaper=%v", runReaper),
+			func(t *testing.T) {
+				suite.Run(t, &ProcessSuite{runReaper: runReaper})
+			},
+		)
 	}
 }

--- a/internal/pkg/dns/dns_test.go
+++ b/internal/pkg/dns/dns_test.go
@@ -29,7 +29,7 @@ import (
 )
 
 func TestDNS(t *testing.T) {
-	goleak.VerifyNone(t)
+	t.Cleanup(func() { goleak.VerifyNone(t) })
 
 	tests := []struct {
 		name         string
@@ -47,7 +47,7 @@ func TestDNS(t *testing.T) {
 		},
 		{
 			name:        "failure",
-			hostname:    "google.com",
+			hostname:    "google-fail.com",
 			nameservers: []string{"242.242.242.242"},
 			errCheck:    check.ErrorContains("i/o timeout"),
 		},
@@ -104,7 +104,7 @@ func TestDNS(t *testing.T) {
 }
 
 func TestDNSEmptyDestinations(t *testing.T) {
-	goleak.VerifyNone(t)
+	t.Cleanup(func() { goleak.VerifyNone(t) })
 
 	stop := newManager(t)
 	defer stop()
@@ -123,7 +123,7 @@ func TestDNSEmptyDestinations(t *testing.T) {
 }
 
 func Test_ServeBackground(t *testing.T) {
-	goleak.VerifyNone(t)
+	t.Cleanup(func() { goleak.VerifyNone(t) })
 
 	m := dns.NewManager(&testReader{}, func(e suture.Event) { t.Log("dns-runners event:", e) }, zaptest.NewLogger(t))
 

